pkg/standalone-utils/contracts/relayer/LidoWrapping.sol

```diff
@@ -0,0 +1,151 @@
+        // to be sourced from outside the relayer, we must first them pull them here.
+        if (sender != address(this)) {
+            require(sender == msg.sender, "Incorrect sender");
+            _pullToken(sender, _stETH, amount);
+        }
+
+        _stETH.approve(address(_wstETH), amount);
+        uint256 result = IwstETH(_wstETH).wrap(amount);
+
+        if (recipient != address(this)) {
+            _wstETH.transfer(recipient, result);
+        }
+
+        if (_isChainedReference(outputReference)) {
+            _setChainedReferenceValue(outputReference, result);
+        }
+    }
+
+    function unwrapWstETH(
+        address sender,
+        address recipient,
+        uint256 amount,
+        uint256 outputReference
+    ) external payable {
+        if (_isChainedReference(amount)) {
+            amount = _getChainedReferenceValue(amount);
+        }
+
+        // The unwrap caller is the implicit sender of tokens, so if the goal is for the tokens
+        // to be sourced from outside the relayer, we must first them pull them here.
+        if (sender != address(this)) {
+            require(sender == msg.sender, "Incorrect sender");
+            _pullToken(sender, _wstETH, amount);
+        }
+
+        uint256 result = _wstETH.unwrap(amount);
+
+        if (recipient != address(this)) {
+            _stETH.transfer(recipient, result);
+        }
+
+        if (_isChainedReference(outputReference)) {
+            _setChainedReferenceValue(outputReference, result);
+        }
+    }
+
+    function stakeETH(
+        address recipient,
+        uint256 amount,
+        uint256 outputReference
+    ) external payable {
+        if (_isChainedReference(amount)) {
+            amount = _getChainedReferenceValue(amount);
+        }
+
+        uint256 result = _stETH.submit{ value: amount }(address(this));
+
+        if (recipient != address(this)) {
+            _stETH.transfer(recipient, result);
+        }
+
+        if (_isChainedReference(outputReference)) {
+            _setChainedReferenceValue(outputReference, result);
+        }
+    }
+
+    function stakeETHAndWrap(
+        address recipient,
+        uint256 amount,
+        uint256 outputReference
+    ) external payable {
+        if (_isChainedReference(amount)) {
+            amount = _getChainedReferenceValue(amount);
+        }
+
+        // The fallback function on the wstETH contract automatically stakes and wraps any ETH which is sent to it.
+        // We can then safely just send the ETH and just have to ensure that the call doesn't revert.
+        payable(address(_wstETH)).sendValue(amount);
+
+        // Note that the previous step would be dangerous should _wstETH be set to the zero address,
+        // however in this scenario the next line will revert, preventing loss of funds.
+
+        // As the wstETH contract doesn't return how much wstETH was minted we must query this separately.
+        uint256 result = _wstETH.getWstETHByStETH(amount);
```

Comment:
(I'm not 100% sure about this, it'd be great if you could double-check)

Reply:
I don't think that's right.

Staking ETH into stETH shouldn't change the exchange rate wstETH:stETH and neither should wrapping and unwrapping stETH (it would be a bad wrapper if doing this affected the exchange rate)

stETH is 1:1 with ETH so when I stake 10% of the current `totalPooledEther` I'm minted 10% of the current total supply of stETH (and so 10% of `_getTotalShares()`).

```
_ethAmount * (_getTotalShares() / totalPooledEther) -> _ethAmount * (1.1 *_getTotalShares()) / (1.1 * totalPooledEther)
```

The extra factors cancel and we have the same exchange rate as before.